### PR TITLE
Add controller layer

### LINKED ISSUE #243: [QUEST] abstract out a 'controller' layer in between routes and models
Right now routes are dealing directly with models.
As a lot of redundancy in page routes and api routes increase, we need a controller layer. (For eg `/api/users/me` and `/users/me` )

What will controllers do ? 
Controllers will supply functions like the following
(written in Typescript to show return types
```ts

function getUsers(): Promise<User[]>
function getUserById(userId: number): Promise<User>
```

Which means we can have routes simply do this - 

```js
// API routes
app.get('/api/users/:id', async (req, res) => {

	const user = await getUserById(req.params.id)
	res.json(user)

})

// Page routes
app.get('/users/:id', async (req, res) => {
	const user = await getUserById(req.params.id)
	res.render('user/id', user)
})
```

We need to start abstracting the most used models into controllers first 

- [ ] users
  - [ ] demographics (including addresses)
- [ ] clients
- [ ] authtokens

### PR DESCRIPTION
**Pull request by Dhroov7**
fixes #243

### CODE REVIEW

File: src/controllers/authTokens.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                token: generator.genNcharAlphaNum(config.AUTH_TOKEN_SIZE),
+                scope: ['*'],
+                explicit: true,
+                clientId: grantCode.clientId,
+                userId: grantCode.userId
+            }
+        })
+    },
+
+    createGrandCode: (clientId,userId) => {
```
Comment (championswimmer) [inline src/controllers/authTokens.js]: grant code, not grand code

**Comment by Dhroov7** [comment]
@championswimmer Done with the changes.

File: src/controllers/client.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        }, {
+            where: {id: clientId}
+        })
+    },
+
+    findAllClients:() => {
+        return models.Client.findAll({})
+    },
+
+    findOneClient:(reqQuery) => {
```
Comment (championswimmer) [inline src/controllers/client.js]: should use variable `clientId`

File: src/controllers/demographics.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    },
+
+    findAllAddress :(userId) => {
+        return models.Address.findAll({
+            where: {'$demographic.userId$': userId},
+            include: [models.Demographic]
+        })
+    },
+
+    findAddress:(Id,userId) => {
```
Comment (championswimmer) [inline src/controllers/demographics.js]: `id` not `Id`

Comment (championswimmer) [inline src/controllers/demographics.js]: `id` not `Id`

**Comment by Dhroov7** [comment]
@championswimmer Done with the changes.

File: src/routers/pages/public.js [hunk 4]
```diff
@@ -7,18 +7,20 @@ const cel = require('connect-ensure-login')
...
-router.get('/signup', cel.ensureNotLoggedIn('/'), function (req, res, next) {
-    Promise.all([
-        models.College.findAll({}), 
-        models.Branch.findAll({})
-    ]).then(function ([colleges, branches]) {
-        res.render('signup', {title: "Signup | OneAuth", colleges:colleges, branches:branches})
-    })
+router.get('/signup', cel.ensureNotLoggedIn('/'), async (req, res, next) => {
+    try{
+        const colleges = await demographicsController.getColleges()
```
Comment (championswimmer) [inline src/routers/pages/public.js]: These two calls can happen parallelly. Via promises they were happening parallely. in the await syntax they are not.

Comment (championswimmer) [reply src/routers/pages/public.js]: Read this to understand more and solve it https://medium.freecodecamp.org/avoiding-the-async-await-hell-c77a0fb71c4c

File: src/routers/pages/user.js [hunk 5]
```diff
@@ -9,67 +9,44 @@ const passutils = require('../../utils/password')
...
-                        models.Company,
-                    ]
-                }
-            ]
-        }).then(function (user) {
-            if (!user) {
-                res.redirect('/login')
+    async (req, res, next) => {
+        try{
+            let includes = []
```
Comment (championswimmer) [inline src/routers/pages/user.js]: why push when you can declare an array in the first place (as it was earlier)

File: src/routers/pages/user.js [hunk 6]
```diff
@@ -9,67 +9,44 @@ const passutils = require('../../utils/password')
...
-            models.Branch.findAll({})
-        ]).then(function ([user, colleges, branches]) {
-            if (!user) {
-                res.redirect('/login')
+    async (req, res, next) => {
+
+        let includes = []
+        includes.push({model:models.Demographic,include:[models.College,models.Branch,models.Company]})
+        try{
+            const user = await userController.getUserById(req.user.id,includes)
```
Comment (championswimmer) [inline src/routers/pages/user.js]: again 3 awaits are serial. can be parallel

**Comment by championswimmer** [comment]
turning this into github magic PR - 

fixes #243

**Pull request closed without merging**